pydantic/_hypothesis_plugin.py

```diff
@@ -0,0 +1,169 @@
+"""
+
+import ipaddress
+import json
+import math
+from typing import cast
+
+import hypothesis.strategies as st
+
+import pydantic
+import pydantic.color
+from pydantic.networks import import_email_validator
+
+# FilePath and DirectoryPath are explicitly unsupported, as we'd have to create
+# them on-disk, and that's unsafe in general without being told *where* to do so.
+
+# Emails
+try:
+    import_email_validator()
+except ImportError:  # pragma: no cover
+    pass
+else:
+    # Note that these strategies deliberately stay away from any tricky Unicode
+    # or other encoding issues; we're just trying to generate *something* valid.
+    st.register_type_strategy(pydantic.EmailStr, st.emails())  # type: ignore[arg-type]
+    st.register_type_strategy(
+        pydantic.NameEmail,
+        st.builds(
+            '{} <{}>'.format,  # type: ignore[arg-type]
+            st.from_regex('[A-Za-z0-9_]+( [A-Za-z0-9_]+){0,5}', fullmatch=True),
+            st.emails(),
+        ),
+    )
+
+# PyObject - dotted names, in this case taken from the math module.
+st.register_type_strategy(
+    pydantic.PyObject,
+    st.sampled_from(
+        [cast(pydantic.PyObject, f'math.{name}') for name in sorted(vars(math)) if not name.startswith('_')]
+    ),
+)
+
+# CSS3 Colors; as name, hex, rgb(a) tuples or strings, or hsl strings
+_color_regexes = (
+    pydantic.color.r_hex_short,
+    pydantic.color.r_hex_long,
+    pydantic.color.r_rgb,
+    pydantic.color.r_rgba,
+    pydantic.color.r_hsl,
+    pydantic.color.r_hsla,
+)
+st.register_type_strategy(
+    pydantic.color.Color,
+    st.one_of(
+        st.sampled_from(sorted(pydantic.color.COLORS_BY_NAME)),
+        st.tuples(
+            st.integers(0, 255),
+            st.integers(0, 255),
+            st.integers(0, 255),
+            st.none() | st.floats(0, 1) | st.floats(0, 100).map('{}%'.format),
+        ),
+        st.from_regex('|'.join(_color_regexes), fullmatch=True),
+    ),
+)
+
+# JSON strings, optionally constrained to a particular type.  We have to register
+# separate strategies for these cases because they're distinct types at runtime.
+st.register_type_strategy(
+    pydantic.Json,
+    st.builds(
+        json.dumps,  # type: ignore[arg-type]
+        st.recursive(
+            base=st.one_of(
+                st.none(),
+                st.booleans(),
+                st.integers(),
+                st.floats(allow_infinity=False, allow_nan=False),
+                st.text(),
+            ),
+            extend=lambda x: st.lists(x) | st.dictionaries(st.text(), x),
+        ),
+        ensure_ascii=st.booleans(),
+        indent=st.none() | st.integers(0, 16),
+        sort_keys=st.booleans(),
+    ),
+)
+
+# Card numbers, valid according to the Luhn algorithm
+
+
+def fix_luhn_digit(card_number: str) -> str:
+    # See https://en.wikipedia.org/wiki/Luhn_algorithm
+    assert card_number[-1] == 'X', 'expected placeholder check digit'
```

Comment:
I'm not quite clear what we're doing with the `X` here? If it's just to do with keeping coverage at 100%, better to just add a pragma.